src/rust/cryptography-x509/src/certificate.rs

```diff
@@ -14,6 +27,52 @@ pub struct Certificate<'a> {
     pub signature: asn1::BitString<'a>,
 }
 
+impl Certificate<'_> {
+    /// Returns the certificate's issuer.
+    pub fn issuer(&self) -> &NameReadable<'_> {
+        self.tbs_cert.issuer.unwrap_read()
+    }
+
+    /// Returns the certificate's subject.
+    pub fn subject(&self) -> &NameReadable<'_> {
+        self.tbs_cert.subject.unwrap_read()
+    }
+
+    /// Returns whether the certificate is "self-issued", whether its
+    /// issuer and subject are the same.
+    pub fn is_self_issued(&self) -> bool {
+        self.issuer() == self.subject()
+    }
+
+    /// Returns an iterable container over the certificate's extension, or
+    /// an error if the extension set contains a duplicate extension.
+    pub fn extensions(&self) -> Result<Extensions<'_>, CertificateError> {
+        self.tbs_cert
+            .extensions()
+            .map_err(CertificateError::DuplicateExtension)
+    }
+
+    /// Returns whether the given extension (by OID) is critical, or
+    /// false if the extension is not present.
+    pub fn extension_is_critical(&self, oid: &asn1::ObjectIdentifier) -> bool {
+        match self.extensions() {
+            Ok(exts) => exts
+                .get_extension(oid)
+                .map(|ext| ext.critical)
+                .unwrap_or(false),
+            Err(_) => false,
+        }
+    }
+
+    /// Returns a specific extension by OID.
+    pub fn extension(
+        &self,
+        oid: &asn1::ObjectIdentifier,
+    ) -> Result<Option<Extension<'_>>, CertificateError> {
+        Ok(self.extensions()?.get_extension(oid))
+    }
```

Comment:
I don't think we want this -- it incurs re-validation of extensions every time.The point of the extensions object is to avoid revalidating things.

Reply:
Makes sense -- I can't think of a nice way to provide this API without either re-iterating or doing something else we probably don't want (like separately caching the extensions here), so I'll drop it.